### PR TITLE
doc: add security hld document

### PR DESCRIPTION
**Pull request by dbkinder**
review and publish ACRN security hld document, also added the oplus
replacement symbol

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

### CODE REVIEW

**Comment by bzhu5** [comment]
David, how to view the document/content in a more human friendly page like html..? thanks
Bing

**Comment by dbkinder** [comment]
@bzhu5 You can get a pretty close approximation by clicking on the VIEW button when looking at the changed files (https://github.com/projectacrn/acrn-hypervisor/blob/d70f6f864ee9126ca2dc925b07177e348ea499d8/doc/developer-guides/security-hld.rst)

GitHub won't render some of the sphinx directives (such as :ref:), and it won't have the final style of the generated HTML, but the GitHub viewer comes close.

File: doc/developer-guides/security-hld.rst [hunk 1]
```diff
@@ -0,0 +1,1081 @@
...
+Platform Security Feature Virtualization and Enablement
+=======================================================
+
+.. note:: This section is under development
+
+This section talks about how the hypervisor enables host CPU features
+(e.g., SGX) and enables platform features (e.g., HECI), to allow guest
+VMs the ability to use those features.
+
+TPM 2.0 Virtualization (vTPMA)
```
Comment (bzhu5) [inline doc/developer-guides/security-hld.rst]: vTPMA  -> vTPM

**Comment by bzhu5** [comment]
it is good to me but just update above vTPMA.

File: doc/developer-guides/security-hld.rst [hunk 2]
```diff
@@ -0,0 +1,1081 @@
...
+secure world must be removed from EPT paging tables of any guest VM,
+even including the SOS VM.
+
+Third-party libraries
+---------------------
+
+All the third-party libraries must be examined before use to verify
+there are no known vulnerabilities in the library source code.
+Typically, the CVE site https://cve.mitre.org/cve/search_cve_list.html
+can be used to search known vulnerabilities.
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/search known/search for known

File: doc/developer-guides/security-hld.rst [hunk 3]
```diff
@@ -0,0 +1,1081 @@
...
+
+.. _platform_root_of_trust:
+
+Platform Root of Trust Key/SEED Derivation
+==========================================
+
+For security reason, each guest VM requires a root key, is used to
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: Missing "which" before "is" here perhaps?

File: doc/developer-guides/security-hld.rst [hunk 4]
```diff
@@ -0,0 +1,1081 @@
...
+   Platform SEED (pSEED) Derivation
+
+As shown in :numref:`security-seed` above, the hypervisor then derives
+multiple child SEEDs for multiple guest VMs. A guest VM must not be able
+to know the SEEDs of any other guest VMs.
+
+The algorithm used in the hypervisor to derive keys is HKDF (HMAC-based
+Extract-and-Expand Key Derivation Function, `RFC5869
+<https://tools.ietf.org/html/rfc5869>`_.  The crypto library `mbedtls
+<https://github.com/ARMmbed/mbedtls>`_ is chosen for project ACRN.
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/is/has been ?

File: doc/developer-guides/security-hld.rst [hunk 5]
```diff
@@ -0,0 +1,1081 @@
...
+#. VMInfo= vm-uuid (from hypervisor configuration file)
+#. theHash=SHA-256
+#. OutSeedLen = 64 in bytes
+#. Guest Dev and User SEED (dvSEED/uvSEED)
+
+   dvSEED = HKDF(theHash, nil, dSEEd, VMInfo\|"devseed", OutSeedLen)
+
+   uvSEED = HKDF(theHash, nil, uSEEd, VMInfo\|"userseed", OutSeedLen
+
+.. _secure_trusty:
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: Should this also be a good section to include a pointer to https://github.com/projectacrn/acrn-hypervisor/blob/master/doc/developer-guides/trusty.rst?

@bzhu5

File: doc/developer-guides/security-hld.rst [hunk 6]
```diff
@@ -0,0 +1,1081 @@
...
+VrKey is the virtual RPMB authentication key used for authentication
+between the SOS DM module and its corresponding UOS secure software.
+Each UOS (if secure storage is supported) has its own VrKey, generated
+randomly when DM process starts, and is securely distributed to UOS
+secure world for each reboot. The rKey is fixed on a specific platform
+unless the eMMC is replaced with another one.
+
+The details of physical RPMB key (rKey) provision are out of scope.  In
+the current project ACRN on APL platform, the rKey is provisioned by
+BIOS (SBL) right after production device ends its manufacture process.
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/manufacture/manufacturing ?

File: doc/developer-guides/security-hld.rst [hunk 7]
```diff
@@ -0,0 +1,1081 @@
...
+
+For each reboot, the BIOS/SBL always retrieves the rKey from CSE FW
+(or generated from a special SEED that is retrieved from CSE FW, refer
+to :ref:`platform_root_of_trust`). The SBL hands this over to the
+ACRN hypervisor, and hypervisor in turn sends to it to the SOS kernel.
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/sends to it /sends it

File: doc/developer-guides/security-hld.rst [hunk 8]
```diff
@@ -0,0 +1,1081 @@
...
+As an example, secure storage virtualization workflow for data write
+access is like this:
+
+#. UOS Secure world (e.g. Trusty) packs the encrypted data and signs it
+   with the vRPMB authentication key (VrKey), and sends the data along
+   with its signature over the RPMB FE driver in UOS non-secure world.
+#. After DM process in SoS receives the data and signature, then the
+   vRPMB module in DM verifies them with the shared secret (vRPMB
+   authentication key, VrKey),
+#. If verification is success, the vRPMB module does data address remap
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/success/successful ?

File: doc/developer-guides/security-hld.rst [hunk 9]
```diff
@@ -0,0 +1,1081 @@
...
+   (remembering that the multiple UOS VMs share a single physical RPMB
+   partition), and forwards those data to SoS kernel. The kernel packs
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/those/the
s/SoS/SOS

File: doc/developer-guides/security-hld.rst [hunk 10]
```diff
@@ -0,0 +1,1081 @@
...
+#. If the verification is successful in eMMC RPMB controller, then the
+   data will be written into storage device.
+
+This work flow of authenticated data read is very similar to this flow
+above, but in reverse order.
+
+Note that there are some security considerations in this design:
+
+#. The rKey protection is very critical in this system. If  it is
+   leaked, an attacker can overwrite the data on RPMB, then which
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: The "then which" sounds odd

File: doc/developer-guides/security-hld.rst [hunk 11]
```diff
@@ -0,0 +1,1081 @@
...
+   violates the "tamper-resistant & anti-replay" capability.
+#. Typically, the vRPMB module in DM process of SOS system can filter
+   data access, preventing one UOS to perform read/write access to the
+   data from another UOS VM. If the vRPMB module in the DM process is
+   compromised, one UOS may also change/overwrite the secure data of
+   other UOS.
+
+Keeping the SOS system as secure as possible is very important goal in
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/is very/is a very

File: doc/developer-guides/security-hld.rst [hunk 12]
```diff
@@ -0,0 +1,1081 @@
...
+------------------------------
+
+On APL platform, Intel |reg| PTT (Platform Trust Technology) implements TPM
+functionalities based on TCG TPM 2.0 industry standard specification.
+PTT exposes TPM hardware interface as CRB (Command Response Buffer)
+defined in the TCG TPM 2.0 spec.
+
+However, in project ACRN, TPM virtualization doesn't assume it is based
+on PTT or discrete TPM; both TPMs (2.0) are supported by design.
+Customers are free to use either PTT or discrete TPM (but not the same
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: s/not the/not at the

**Comment by dbkinder** [comment]
Thanks @gvancuts, typos are fixed (it was a late night :)

**Comment by gvancuts** [comment]
> Thanks @gvancuts, typos are fixed (it was a late night :)

Thanks! I reckoned you must have been a little tired, there were more typos than usual :)

**Comment by dbkinder** [comment]
@gvancuts close enough to approve? or do you want @bzhu5 to address comments before we publish this version?

File: doc/developer-guides/security-hld.rst [hunk 13]
```diff
@@ -0,0 +1,1081 @@
...
+hypervisor. Secure world has a higher "privilege level" than non-secure
+world, for example, the secure world can access the non-secure world's
+physical memory but not vice-versa. This document discuss how this
+security works and why it is required.
+
+Note that the SOS is a closed system, and only OEM modules and
+applications can be installed run within the SOS system. The other guest
+VMs (UOSs) are open system, and, for example, an application "store"
+could be available to allow the end user to download, install, and run
+applications from 3rd-party developers.
```
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: sure, you can.

File: doc/developer-guides/security-hld.rst [hunk 14]
```diff
@@ -0,0 +1,1081 @@
...
+
+
+Background
+**********
+
+The ACRN hypervisor is a type-1 hypervisor, built for running multiple
+guest OS instances, typical of an automotive infotainment system, on a
+single Intel Apollo Lake-I SoC platform. See :numref:`security-ACRN`.
+
+.. figure:: images/security-image1.png
```
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: i never worked on UEFI platform in any product, sorry, i cannot answer you at this moment.

File: doc/developer-guides/security-hld.rst [hunk 15]
```diff
@@ -0,0 +1,1081 @@
...
+======================
+
+The security of the entire system built on top of the ACRN hypervisor
+depends on the security from platform boot to UOS launching. Each layer
+or module must verify the security of the next layer or module before
+transferring control to it. Verification could be checking a
+cryptographic signature on the executable of the next step before it is
+launched.
+
+Note that measured boot is not currently supported for ACRN and its
```
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: good question. this comes from TCG (trustedcomputinggroup.org), here it is a doc.
https://trustedcomputinggroup.org/wp-content/uploads/DRTM-Specification-Overview_June2013.pdf

File: doc/developer-guides/security-hld.rst [hunk 16]
```diff
@@ -0,0 +1,1081 @@
...
+Similarly, virtual cold/warm reboot is also supported. How to implement
+virtual power life cycle management is out of scope in this document.
+
+This subsection is intended to describe the security issues for those
+power cycles.
+
+UOS Power On and Shutdown
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Based on current UOS work flow design, the memory of guest VM (UOS) is
```
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: maybe we need to revise it by just removing the first sentences

File: doc/developer-guides/security-hld.rst (hunk 4 above)
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: ok.

File: doc/developer-guides/security-hld.rst (hunk 5 above)
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: but this is done internally in hypervisor when hcall_initialize_trusty() is called.

File: doc/developer-guides/security-hld.rst (hunk 6 above)
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: correct:)

File: doc/developer-guides/security-hld.rst (hunk 7 above)
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: correct:)

File: doc/developer-guides/security-hld.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: this is also a good reference: 
Survey of Boot security technologies 
https://firmwaresecurity.com/2015/07/29/survey-of-boot-security-technologies/

**Comment by dbkinder** [comment]
@gvancuts do we get an approved review from you yet so we can publish?

File: doc/developer-guides/security-hld.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: OK, thanks for the confirmation.

Is this applicable to the UP2 board (based on APL) when running SBL?

[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: I meant that the word "kernel" in the picture is missing the final "l".

File: doc/developer-guides/security-hld.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: I cannot easily change the text that you submitted in this PR. Can you update this?

(If @dbkinder wants to update this in a follow-up PR, I'm fine too).

File: doc/developer-guides/security-hld.rst (hunk 5 above)
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: I do not understand your previous comment?

I was suggesting that we add a link to this other document we have that talks about Trusty.

File: doc/developer-guides/security-hld.rst (hunk 7 above)
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: Can someone update this text?

File: doc/developer-guides/security-hld.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (gvancuts) [reply doc/developer-guides/security-hld.rst]: Scrap my last comment, it was @dbkinder who submitted this PR. Can you update the text @dbkinder ?

File: doc/developer-guides/security-hld.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (dbkinder) [reply doc/developer-guides/security-hld.rst]: fixed

**Comment by dbkinder** [comment]
@gvancuts That should address all your comments (image1 fix too)

File: doc/developer-guides/security-hld.rst [hunk 17]
```diff
@@ -0,0 +1,1079 @@
...
+##########################
+
+.. primary author: Bing Zhu
+   contributor: Yadong Qi
+
+Introduction
+************
+
+This document describes the ACRN security high level design on Apollo
+Lake SoCs, including information about:.
```
Comment (gvancuts) [inline doc/developer-guides/security-hld.rst]: There is a dot right after the semi-colon, is that intended?

Comment (hwang37) [reply doc/developer-guides/security-hld.rst]: @dbkinder @bzhu5 to address the comment

Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: @dbkinder you can remove that "dot", thank you

File: doc/developer-guides/security-hld.rst (hunk 14 above)
[the remark this answers is not in the document]
Comment (bzhu5) [reply doc/developer-guides/security-hld.rst]: i believe yes, if SBL is used.

**Pull request merged**